crypto/zkp/primitives/commitments/damgard-fujisaki_positive.go

```diff
@@ -0,0 +1,202 @@
+ * limitations under the License.
+ *
+ */
+
+package commitmentzkp
+
+import (
+	"math/big"
+
+	"fmt"
+
+	"github.com/xlab-si/emmy/crypto/commitments"
+	"github.com/xlab-si/emmy/crypto/common"
+)
+
+// DFCommitmentPositiveProver proves that the commitment hides the positive number. Given c,
+// prove that c = g^x * h^r (mod n) where x >= 0.
+type DFCommitmentPositiveProver struct {
+	squareProvers []*DFCommitmentSquareProver
+}
+
+func NewDFCommitmentPositiveProver(committer *commitments.DamgardFujisakiCommitter,
+	x, r *big.Int, challengeSpaceSize int) (*DFCommitmentPositiveProver, []*big.Int, []*big.Int, error) {
+
+	// x can be written (if positive) as x = x0^2 + x1^2 + x2^2 + x3^2.
+	// We create committers which hold c0 = g^(x0^2) * h^r0, c1 = g^(x1^2) * h^r1,
+	// c2 = g^(x2^2) * h^r2, c3 = g^(x3^2) * h^r3 and where r = r0 + r1 + r2 + r3.
+	// We then prove that c0, c1, c2, c3 contains squares and verifier checks that c = c0*c1*c2*c3.
+
+	roots, err := common.LipmaaDecomposition(x)
+	if err != nil {
+		return nil, []*big.Int(nil), []*big.Int(nil),
+			fmt.Errorf("error when doing Limpaa decomposition")
+	}
+	numOfRoots := len(roots)
+
+	// find r0, r1, r2, r3 such that r0 + r1 + r2 + r3 = r
+
+	rIsNegative := false
+	rAbs := new(big.Int).Abs(r)
+	if rAbs.Cmp(r) != 0 {
+		rIsNegative = true
+	}
+
+	boundary := new(big.Int)
+	boundary.Set(rAbs)
+
+	var rs []*big.Int
+	for i := 0; i < numOfRoots; i++ {
+		currR := common.GetRandomInt(boundary)
+		if i < numOfRoots-1 {
+			rs = append(rs, currR)
+			boundary.Sub(boundary, currR)
+		} else {
+			rs = append(rs, boundary)
+		}
+	}
+
+	if rIsNegative {
+		for i := 0; i < numOfRoots; i++ {
+			rs[i].Neg(rs[i])
+		}
+	}
+
+	var committers []*commitments.DamgardFujisakiCommitter
+	var commitmentsToSquares []*big.Int
+	for i := 0; i < numOfRoots; i++ {
+		committer := commitments.NewDamgardFujisakiCommitter(committer.QRSpecialRSA.N,
+			committer.H, committer.G, committer.T, committer.K)
+		square := new(big.Int).Mul(roots[i], roots[i])
+		commitment, err := committer.GetCommitMsgWithGivenR(square, rs[i])
+		commitmentsToSquares = append(commitmentsToSquares, commitment)
+		if err != nil {
+			return nil, []*big.Int(nil), []*big.Int(nil),
+				fmt.Errorf("error when creating commit msg")
+		}
+		committers = append(committers, committer)
+	}
+
+	var bases []*big.Int
+	var squareProvers []*DFCommitmentSquareProver
+	for i := 0; i < numOfRoots; i++ {
+		prover, c1, err := NewDFCommitmentSquareProver(committers[i], roots[i], challengeSpaceSize)
+		if err != nil {
+			return nil, []*big.Int(nil), []*big.Int(nil),
+				fmt.Errorf("Error in instantiating DFCommitmentSquareProver")
+		}
+		bases = append(bases, c1)
+		squareProvers = append(squareProvers, prover)
+	}
+
+	return &DFCommitmentPositiveProver{
+		squareProvers: squareProvers,
+	}, bases, commitmentsToSquares, nil
+}
+
+func (prover *DFCommitmentPositiveProver) GetProofRandomData() []*big.Int {
```

Comment:
Please use a short name for method receivers, e.g.  `func (p *DFCommitmentPositiveProver)` instead of `func (prover *DFCommitmentPositiveProver)`. Apply this change to all provers and receivers in this PR. Read more about it [here](https://github.com/golang/go/wiki/CodeReviewComments#receiver-names).